### PR TITLE
Introduce #[pallet::call_index] attribute to dispatchables

### LINKED ISSUE #10408: Allow explicit call ordering within a pallet
Similar to explicit ordering in the `construct_runtime` macro, we could annotate calls within a pallet with an index. This would have a few advantages, namely the ability to remove a function without affecting the indices of other calls.

This could be useful in pallet versioning and transaction version, where a pallet's version would need to increment when there is a breaking change to these indices/function signatures. By leaving call indices empty on function removal, wallets/signers don't need to worry about signing an unintended call as it would be null.

Transaction version could either increment on pallet version changes or be reserved for changes in the serialization (e.g. signed extensions).

### PR DESCRIPTION
**Pull request by KiChjang**
Fixes #10408.

Per title, this PR adds a new attribute `#[pallet::call_index($idx)]` to dispatchable functions, which can be used to fix the codec order of the corresponding enum variant in the `Call` enum.

### CODE REVIEW

**Comment by KiChjang** [comment]
There isn't any logic to check whether an index is duplicated, perhaps I should first include that.

**Comment by kianenigma** [comment]
> There isn't any logic to check whether an index is duplicated, perhaps I should first include that.

Wouldn't the rust compiler already raise an error if you do that? If so, you mean a better error message, rather than letting the user deal with the rustc error? 


UPDATE: having a peek at the code, I don't see where the new value is used. Eventually, the functions that are annotated with `#pallet::call` are used to generate `enum Call { Fn1(..), Fn2(..) }`. We should just change this to `enum Call { Fn1(..) = x, Fn2(..) = y }` and let rustc handle the errors for us as well. I belive this is how the `construct_runtime` also works.

File: frame/support/procedural/src/pallet/parse/call.rs [hunk 1]
```diff
@@ -182,18 +198,42 @@ impl CallDef {
...
+					return Err(syn::Error::new(method.sig.span(), msg))
+				}
+				let call_index = call_idx_attrs.pop().map(|attr| match attr {
+					FunctionAttr::CallIndex(idx) => idx,
+					_ => unreachable!("checked during creation of the let binding"),
+				});
+
+				let final_index =
+					call_index.unwrap_or_else(|| last_index.map(|idx| idx + 1).unwrap_or(0));
+				last_index = Some(final_index);
```
Comment (kianenigma) [inline frame/support/procedural/src/pallet/parse/call.rs:236]: I don't see this being used henceforth, why do we assign to it?

**Comment by KiChjang** [comment]
@kianenigma https://github.com/paritytech/substrate/issues/10408#issuecomment-1120148386 is why we can't assign discriminants to non-unit enum variants. In short, it's an unsupported Rust feature.

**Comment by KiChjang** [comment]
`construct_runtime` in fact uses `#[codec(index = $idx)]` as well, instead of enum discriminants.

File: frame/support/procedural/src/pallet/parse/call.rs (hunk 1 above)
Comment (KiChjang) [reply frame/support/procedural/src/pallet/parse/call.rs:236]: It's being used during expansion to fill up the value in the `#[codec(index = $idx)]` attribute.

File: frame/support/procedural/src/pallet/expand/call.rs [hunk 2]
```diff
@@ -177,6 +178,7 @@ pub fn expand_call(def: &mut Def) -> proc_macro2::TokenStream {
 			),
 			#(
 				#( #[doc = #fn_doc] )*
+				#[codec(index = #call_indices)]
```
Comment (shawntabrizi) [inline frame/support/procedural/src/pallet/expand/call.rs:181]:
```suggestion
				#[codec(index = #call_indices)]
```

intentionally plural?

File: frame/support/src/lib.rs [hunk 3]
```diff
@@ -1645,6 +1645,9 @@ pub mod pallet_prelude {
 /// used using `#[pallet::compact]`, function must return `DispatchResultWithPostInfo` or
 /// `DispatchResult`.
 ///
+/// Each dispatchable may also be annotated with the `#[pallet::call_index($idx)]` attribute,
+/// which defines and sets the codec index for the dispatchable function in the `Call` enum.
```
Comment (shawntabrizi) [inline frame/support/src/lib.rs:1649]: can you briefly add a description here of what happens if you use the call index in some placed, but not in others?

Can we make it so that call index must be used everywhere or nowhere?

**Review by shawntabrizi — APPROVED** [review]

**Comment by shawntabrizi** [comment]
This looks good to me as is, however I am coming to the conclusion that we may want to handle calls differently in the future.

What if rather than having the call enum be generated by including a bunch of functions inside `#[pallet::call]`, we instead start by the user simply creating an explicit call enum with just names.

Then we would implement traits on top of each enum, for example `Dispatchable` which then maps each enum variant to the function it executes, `WeightInfo` which maps each enum variant to the weight function, and even things like `PreExecute`, a function or set of checks which is executed before the main function, `PostExecute`, and whatever other million of things we want.

Basically, by hiding the call enum from the developer, we have not simplified things much, but have made extending calls much less flexible.

Also, this would allow us to have call function definitions in completely different files, and stuff like that...

just an idea

File: frame/support/src/lib.rs (hunk 3 above)
Comment (shawntabrizi) [reply frame/support/src/lib.rs:1649]: Honestly, probably call index should just be mandatory. We could allow users to ignore it if they use [`pallet::dev_mode`](https://github.com/paritytech/substrate/issues/8965)

Comment (kianenigma) [reply frame/support/src/lib.rs:1649]: In the spirit of keeping the FRAME boilerplate minimal, I don't think mandating call-indices should be mandatory. I like how it works in `construct_runtime` and find it intuitive: You have the default order, and you can overwrite anything that you want. Duplicates will be detected at compile time.

**Review by kianenigma — APPROVED** [review]
Thanks! cc @joepetrowski

**Comment by KiChjang** [comment]
bot merge
